src/liblock/lockmail.1
.\"  <!-- $Id: lockmail.1,v 1.1.1.1 2003/05/07 02:14:27 lfan Exp $ -->
.\"  <!-- Copyright 2002 Double Precision, Inc.  See COPYING for -->
.\"  <!-- distribution information. -->
.\" This manpage has been automatically generated by docbook2man 
.\" from a DocBook document.  This tool can be found at:
.\" <http://shell.ipoline.com/~elmert/comp/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "LOCKMAIL" "1" "12 December 2002" "Double Precision, Inc." ""
.SH NAME
lockmail \- create mail lock files
.SH SYNOPSIS

\fBlockmail\fR [ \fB-r\fR ] [ \fB-t \fItimeout\fB\fR ] \fB\fIlockfile\fB\fR \fB\fIprogram\fB\fR [ \fBargument\fR\fI ...\fR ]

.SH "DESCRIPTION"
.PP
\fBlockmail\fR is a helper utility for working with mailbox
files.
Mailbox files must be locked to prevent other applications from modifying the
mailbox at the same time.
Different system use different locking conventions.
\fBlockmail\fR uses two of the most common locking mechanisms
in use, which should work reliably on most systems.
.PP
\fIlockfile\fR is the pathname to an existing mailbox
file.
By default, \fBlockmail\fR tries to lock the mailbox every
five seconds (if the mailbox is already locked), and will give up after
three minutes.
After the mailbox is succesfully locked, \fBlockmail\fR runs
\fIprogram\fR as a child process, with any optional
\fIargument\fRs.
When \fIprogram\fR terminates, \fBlockmail\fR
removes the mailbox lock, and terminates itself.
.SH "OPTIONS"
.TP
\fB-r\fR
If a regular lock fails, try a read-only lock.
Use this option to lock mailbox files in a read-only directory.
.TP
\fB-t \fItimeout\fB\fR
If the lock attempt fails, try again for up to
\fItimeout\fR seconds.
The actual timeout is rounded up to the next five second interval
(a lock attempt is tried every five seconds).
.SH "DESCRIPTION"
.PP
This section briefly describes the locking mechanism used by
\fBlockmail\fR.
\fBlockmail\fR uses three different locking conventions in
order to maximize compatibility with other mail software:
C-Client folder locks, dot-locks, and file locks.
.SS "C-CLIENT FOLDER LOCKS"
.PP
Mail software based on the C-Client library creates
lock files named
\fI/tmp/.dddddd.iiiiii\fR.
Here, \fIdddddd\fR and \fIiiiiii\fR
are the device number and the inode number of the mailbox file
(the \fBst_dev\fR and \fBst_ino\fR
fields in the inode), in hexadecimal.
If the process ID saved in the C-Client folder lock file is not valid,
\fBlockmail\fR concludes that it's a stale lock file, and
will remove it.
.sp
.RS
.B "Note:"
.PP
A race condition exists where a C-Client process is
killed after it creates a lock file, but before saving its process ID in the
lock file.
The race window is very small, but it exists.
The C-Client library does not appear to ever clear out
the lock file.
.PP
\fBlockmail\fR
attempts to resolve this race condition by deleting zero-length lock files
that are at least five minutes old.
.RE
.SS "DOT-LOCKS"
.PP
\fBlockmail\fR
also creates, and honors dot-lock files.
Dot-lock files are first created as temporary files, then linked to
\fIlockfile.lock\fR.
The link operation fails if the dot-lock file already exists.
\fBlockmail\fR
uses an enhanced method of dot-locking, where its process ID, and the name
of the server where \fBlockmail\fR is running is also saved
in its dot-lock file.
If the operation fails due to an existing dot-lock file that was created
by another \fBlockmail\fR process on the same server, and the
process ID no longer exists, this stale dot-lock file is removed immediately.
In all other situations a dot-lock file older than five minutes is considered
stale, and removed.
.sp
.RS
.B "Note:"
A failure to create a dot-lock file is silently ignored if the reason for
the failure is because
\fBlockmail\fR
does not have the write permission in the dot-lock file's directory.
The incoming mail spool directory (usually
\fI/var/spool/mail\fR)
typically does not have global write permissions, so the attempt to
create the dot-lock file in the spool directory will fail, and
\fBlockmail\fR
will be content with using file-locking only.
.RE
.SS "FILE LOCKS"
.PP
The final locking mechanism
\fBlockmail\fR
uses is the operating system's file locking facility.
If
\fBlockmail\fR
fails to obtain all three locks,
\fBlockmail\fR
will sleep for five seconds and try again.
The only exception is a failure to create a dot-lock because of no write
access to the dot-lock file's directory, which is ignored.
If
\fBlockmail\fR
still fails to obtain all required locks in the amount of time specified
by the \fB-t\fR option (or its default value),
\fBlockmail\fR will terminate with the
EX_TEMPFAIL exit code.
.PP
\fBlockmail\fR
runs \fIprogram\fR after obtaining the last file
lock, waits until \fIprogram\fR terminates, and
releases all locks.
\fIprogram\fR must terminate before any of the locks
obtained by \fBlockmail\fR expire, and are considered stale.
\fBlockmail\fR will then terminate with the same exit code
as \fIprogram\fR.
.SH "EXIT STATUS"
.PP
\fBlockmail\fR terminates with the same exit status as
\fIprogram\fR
\fBlockmail\fR terminates with the EX_TEMPFAIL
exit status if it was unable to obtain a lock, or if
\fIprogram\fR
was killed by a signal.
.SH "SEE ALSO"
.PP
\fBmaildrop\fR(1),
\fBsendmail\fR(8).


src/liblock/lockmail.c
/*
** Copyright 2002 Double Precision, Inc.  See COPYING for
** distribution information.
*/

#include	"config.h"
#include	"liblock.h"
#include	"mail.h"
#include	<stdio.h>
#include	<stdlib.h>
#include	<string.h>
#include	<unistd.h>
#include	<errno.h>
#include	<signal.h>
#include	<sys/types.h>
#include	<sys/stat.h>
#if HAVE_SYS_WAIT_H
#include <sys/wait.h>
#endif
#ifdef HAVE_FCNTL_H
#include <fcntl.h>
#endif
#ifndef WEXITSTATUS
#define WEXITSTATUS(stat_val) ((unsigned)(stat_val) >> 8)
#endif
#ifndef WIFEXITED
#define WIFEXITED(stat_val) (((stat_val) & 255) == 0)
#endif

#if HAVE_SYSEXITS_H
#include <sysexits.h>
#endif

#ifndef EX_TEMPFAIL
#define EX_TEMPFAIL 75
#endif

static const char rcsid[]="$Id: lockmail.c,v 1.1.1.1 2003/05/07 02:14:27 lfan Exp $";

#define NTRIES_DEFAULT	60
#define DELAY	5

static int sig=0;

static int catch(int signum)
{
	sig=1;
	signal(SIGHUP, (RETSIGTYPE (*)(int))catch);
	signal(SIGTERM, (RETSIGTYPE (*)(int))catch);
	signal(SIGINT, (RETSIGTYPE (*)(int))catch);
	return 0;
}

static int caught()
{
	signal(SIGHUP, SIG_DFL);
	signal(SIGTERM, SIG_DFL);
	signal(SIGINT, SIG_DFL);
	return sig;
}

int main(int argc, char **argv)
{
	char **argvec;
	int n;
	int fd;
	pid_t pid, pid2;
	int waitstat;
	struct ll_mail *p;
	int ntries=NTRIES_DEFAULT;
	int readonly=0;
	int optchar;

	while ((optchar=getopt(argc, argv, "+rt:")) != -1)
		switch (optchar) {
		case 'r':
			readonly=1;
			break;
		case 't':
			ntries=atoi(optarg);
			if (ntries < 0)
			{
				fprintf(stderr, "%s: invalid argument to -t\n",
					argv[0]);
				exit(EX_TEMPFAIL);
			}
			ntries= (ntries / DELAY) + 1;
			break;
		default:
			exit(1);
		}

	if (argc - optind < 2)
	{
		fprintf(stderr, "Usage: %s [-r] [-t time] mailfile program [args]...\n",
			argv[0]);
		exit(1);
	}

	if ((argvec=malloc(sizeof(char *) * (argc - optind + 1))) == NULL)
	{
		perror("malloc");
		exit (1);
	}

	for (n=optind+1; n<argc; n++)
		argvec[n-optind - 1]=argv[n];

	argvec[n-optind-1]=NULL;

	/* Create the mail file, if it doesn't exist */

	if ((n=open(argv[optind], O_RDWR|O_CREAT, 0600)) >= 0)
		close(n);

	signal(SIGUSR2, SIG_IGN);
	signal(SIGCHLD, SIG_DFL);

	p=NULL;

	for (n=0; n<ntries; sleep(DELAY), n++)
	{
		if (p)
			ll_mail_free(p);

		if ((p=ll_mail_alloc(argv[optind])) == NULL)
		{
			perror("malloc");
			exit(1);
		}

		signal(SIGHUP, (RETSIGTYPE (*)(int))catch);
		signal(SIGTERM, (RETSIGTYPE (*)(int))catch);
		signal(SIGINT, (RETSIGTYPE (*)(int))catch);

		if (ll_mail_lock(p) < 0)
		{
			if (errno == ENOENT)
				break; /* Mail file gone? */
			if (caught())
				break;
			continue;
		}

		if ((fd=ll_mail_open(p)) < 0)
		{
			if (!readonly || (fd=ll_mail_open_ro(p)) < 0)
			{
				if (caught())
					break;
				continue;
			}
		}

		if ((pid=fork()) < 0)
		{
			perror("fork");
			exit(1);
		}

		if (pid == 0)
		{
			(void)caught();
			execvp(argvec[0], argvec);

			perror(argvec[0]);
			exit(1);
		}

		while ((pid2=wait(&waitstat)) != pid)
			;

		ll_mail_free(p);

		if (WIFEXITED(waitstat))
			exit(WEXITSTATUS(waitstat));
		exit(EX_TEMPFAIL);
	}
	if (p)
		ll_mail_free(p);
	exit(EX_TEMPFAIL);
	return (0);
}

